data_utils/ds_addmult_mod10_helpers.py

```diff
@@ -155,20 +207,36 @@ def eval_callback_mod10(model, in_vocab, split, datasets, eval_batch_size=32, du
             label = batch['labels'].long().to(device)
 
             # Get the model's predictions
-            outputs = model(input, input_len, None, None, None)
-            _, predicted_class = torch.max(outputs, dim=1)
-
-            # Update accuracy counters
-            total_count += label.size(0)
-            correct_count += (predicted_class == label).sum().item()
-
-            if (dump):
-                strings = batch['string']
-                label = label.cpu().numpy()
-                predicted_class = predicted_class.cpu().numpy()
-                for idx in range(len(label)):
-                    if (label[idx] != predicted_class[idx]):
-                        dump_file.write(strings[idx] + '\n')
+            if has_token_labels: # TransformerLM
+                prefix_sos = (torch.ones((input.shape[0],1)).long() * model.encoder_sos).to(device)
+                input = torch.cat((prefix_sos, input), dim=1) # add SOS token
+                input_len += 1
+
+                outputs = model(input, input_len)
+                _, preds = torch.max(outputs['data'], dim=2)
+
+                labels_cmp = torch.cat((label.T, prefix_sos), dim=1)
+                labels_match = (preds == labels_cmp) * (input != 0) * (labels_cmp != 0)
+
+                # Update accuracy counters
+                total_count += batch['labels_len'].sum().item()
+                correct_count += labels_match.sum().item()
+
+            else: # EncDec
+                outputs = model(input, input_len, None, None, None)
+                _, predicted_class = torch.max(outputs, dim=1)
+                # Update accuracy counters
+                total_count += label.size(0)
+                correct_count += (predicted_class == label).sum().item()
+
+                if (dump):
+                    with open(dump_file_path, 'w') as dump_file:
```

Comment:
1. This will overwrite the file each time we pass through a batch, leaving only the data gathered from the batch from the last callback we executed. This should be pulled out of this loop, and only write once data is gathered (fine to keep in same function though).

2. We can then add mismatches from 210-223 in the `if has_token_labels` function (if it is easy to do so!), and allow dump functionality for both EncDec and TransformerLM

Else, please ensure an error is raised on line 211:
```
raise NotImplementedError("eval_callback_mod10 is unable to dump to file when 'has_token_labels' is set")
```

Reply:
- @ananjan-nandi-9 - Is `dump()` an evolutionary dead end, can it be pruned?

- This was a rough sketch of a feature carried forward from Ananjan's `eval_callback_mod10_lm`, to provide interface compatibility
- It's never been used in my code, doesn't seem to be called anywhere
- There's no point carrying design overhead for one-off failed experiments
- I've made the suggested fix, but really if we're not using it we should just delete

Reply:
Thank you for adding suggested fix, going to resolve this, but we can continue the conversation here.

Reply:
Noted, the intention with this was to dump mismatches for some analysis I was doing. I think this can be removed.